// @gravity-ui/uikit/build/esm/components/tabs/hooks/useTab.js
import * as React from 'react';
import { KeyCode } from "../../../constants.js";
import { filterDOMProps } from "../../utils/filterDOMProps.js";
import { TAB_DATA_ATTRIBUTE, bTab } from "../constants.js";
import { TabContext } from "../contexts/TabContext.js";
export function useTab(tabProps) {
    const tabContext = React.useContext(TabContext);
    if (!tabContext) {
        throw new Error('<Tab> must be used within <TabList>');
    }
    const currentValue = tabContext.value;
    const tabId = `${tabContext.id}:t:${tabProps.value}`;
    const panelId = tabContext.isProvider ? `${tabContext.id}:p:${tabProps.value}` : undefined;
    const isSelected = currentValue === tabProps.value;
    const isDisabled = tabProps.disabled;
    const isFocused = tabContext.isFocused;
    const onClick = () => {
        if (!tabProps.disabled) {
            tabContext.onUpdate?.(tabProps.value);
        }
    };
    const onKeyDown = (event) => {
        if ((event.key === KeyCode.SPACEBAR || event.key === KeyCode.ENTER) && !tabProps.disabled) {
            tabContext.onUpdate?.(tabProps.value);
        }
    };
    return {
        role: 'tab',
        'aria-selected': isSelected,
        'aria-disabled': isDisabled,
        'aria-controls': panelId,
        id: tabId,
        tabIndex: isSelected && !isDisabled && !isFocused ? 0 : -1,
        onClick,
        onKeyDown,
        title: tabProps.title,
        style: tabProps.style,
        className: bTab({ active: isSelected, disabled: isDisabled }, tabProps.className),
        'data-qa': tabProps.qa,
        [TAB_DATA_ATTRIBUTE]: tabProps.value,
        ...filterDOMProps(tabProps, { labelable: true }),
    };
}
